esp32/machine_timer.c

```diff
@@ -0,0 +1,193 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdint.h>
+#include <stdio.h>
+
+#include "driver/timer.h"
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "modmachine.h"
+
+#define TIMER_INTR_SEL TIMER_INTR_LEVEL
+#define TIMER_DIVIDER  40000
+#define TIMER_SCALE    (TIMER_BASE_CLK / TIMER_DIVIDER)
+
+#define TIMER_FLAGS    0
+
+typedef struct _machine_timer_obj_t {
+    mp_obj_base_t base;
+    mp_uint_t group;
+    mp_uint_t index;
+
+    mp_uint_t repeat;
+    mp_uint_t period;
+
+    mp_obj_t callback;
+
+    intr_handle_t handle;
+} machine_timer_obj_t;
+
+const mp_obj_type_t machine_timer_type;
+
+STATIC esp_err_t machine_try(esp_err_t code) {
+    if (code) {
+        mp_raise_OSError(code);
+    }
+
+    return code;
+}
+
+STATIC void machine_timer_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_timer_obj_t *self = self_in;
+
+    timer_config_t config;
+    mp_printf(print, "Timer(%p) [", self);
+
+    timer_get_config(self->group, self->index, &config);
+
+    mp_printf(print, "alarm_en=%d ", config.alarm_en);
+    mp_printf(print, "auto_reload=%d ", config.auto_reload);
+    mp_printf(print, "counter_en=%d ", config.counter_en);
+    mp_printf(print, "]");
+}
+
+STATIC mp_obj_t machine_timer_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+    machine_timer_obj_t *self = m_new_obj(machine_timer_obj_t);
+    self->base.type = &machine_timer_type;
+
+    self->group = (mp_obj_get_int(args[0]) >> 1) & 1;
+    self->index = mp_obj_get_int(args[0]) & 1;
+
+    return self;
+}
+
+STATIC void machine_timer_disable(machine_timer_obj_t *self) {
+    if (self->handle) {
+        timer_pause(self->group, self->index);
+        esp_intr_free(self->handle);
+        self->handle = NULL;
+    }
+}
+
+STATIC void IRAM_ATTR machine_timer_isr(void *self_in) {
```

Comment:
I checked the address (or rather looked at PC when the isr generated a guru meditation), and it's always seems to be in the 0x40092BE0 block irrespective of `IRAM_ATTR`. I was initially nervous that omitting it would be bad, but was probably being overly sensitive.